docs/path-resolution.rst

```diff
@@ -0,0 +1,690 @@
+The normalization rules are the same as for UNIX paths, namely:
+
+- All the ``./`` segments are removed.
+- Every ``../``  segment backtracks one level up in the hierarchy.
+- Multiple slashes are squashed into a single one.
+
+Example:
+
+.. code-block:: solidity
+    :caption: lib/contract.sol
+
+    import "./util/./util.sol";         // source unit name: lib/util/util.sol
+    import "./util//util.sol";          // source unit name: lib/util/util.sol
+    import "../util/../array/util.sol"; // source unit name: array/util.sol
+
+.. warning::
+
+    The root of the virtual filesystem is an empty path, not ``/``.
+    This matters when the ``../`` segments go beyond the root.
+    In UNIX paths such segments are ignored and for example ``/../../`` is
+    equivalent to just ``/``.
+    In the virtual filesystem the rule similar but the result is an empty path instead.
+
+    .. code-block:: solidity
+        :caption: /project/lib/contract.sol
+
+        import "../util.sol";          // source unit name: /project/util.sol
+        import "../../util.sol";       // source unit name: /util.sol
+        import "../../../util.sol";    // source unit name: util.sol
+        import "../../../../util.sol"; // source unit name: util.sol
+
+.. note::
+
+    The parent source unit name is **not** normalized.
+    This ensures that relative imports work properly when the importing file is identified with an URL:
+
+    .. code-block:: solidity
+        :caption: https://example.com/contract.sol
+
+        import "./token.sol"; // source unit name: https://example.com/token.sol
+
+    If the parent source unit name were to be normalized, the name would become
+    ``https:/example.com/token.sol`` which is not a valid URL.
+
+.. warning::
+
+    The ``./`` and ``../`` segments in the parent source unit name have no special meaning.
+
+    .. code-block:: solidity
+        :caption: ../lib/../lib/math.sol
+
+        import "./util.sol" as util;    // source unit name: ../lib/../lib/util.sol
+        import "../token.sol" as token; // source unit name: ../lib/../../token.sol
+
+    This may lead to surprising results in corner cases.
+    For example they can get canceled by ``../`` segments in the import path:
+
+    .. code-block:: solidity
+        :caption: /project/./lib/contract.sol
+
+        import "../util.sol";       // source unit name: /project/./util.sol
+        import "../../util.sol";    // source unit name: /project/util.sol
+        import "../../../util.sol"; // source unit name: /util.sol
+
+.. note::
+
+    The use of relative imports containing leading ``../`` segments is not recommended.
+    The same effect can be achieved in a more reliable way by using direct imports with
+    :ref:`base path <base-path>` and :ref:`import remapping <import-remapping>`.
+
+.. index:: ! base path
+.. _base-path:
+
+Base Path
+=========
+
+Base path is the directory that contains all the files belonging to the project being compiled.
```

Comment:
I'm not sure this is a good definition.

Reply:
Reworded.

I think this is basically what it boils down to in the current implementation - if you use non-empty base path everything has to sit in a single directory.